nvflare/apis/utils/reliable_message.py

```diff
@@ -0,0 +1,369 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import concurrent.futures
+import threading
+import time
+import uuid
+
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import ReservedHeaderKey, ReturnCode, Shareable, make_reply
+from nvflare.apis.signal import Signal
+
+# Operation Types
+OP_REQUEST = "req"
+OP_QUERY = "query"
+OP_REPLY = "reply"
+
+# Reliable Message headers
+HEADER_OP = "rm.op"
+HEADER_TOPIC = "rm.topic"
+HEADER_TX = "rm.tx"
+HEADER_TIMEOUT = "rm.timeout"
+HEADER_STATUS = "rm.status"
+
+# Status
+STATUS_IN_PROCESS = "in_process"
+STATUS_IN_REPLY = "in_reply"
+STATUS_NOT_RECEIVED = "not_received"
+STATUS_REPLIED = "replied"
+STATUS_ABORTED = "aborted"
+
+# Topics for Reliable Message
+TOPIC_RELIABLE_REQUEST = "RM.RELIABLE_REQUEST"
+TOPIC_RELIABLE_REPLY = "RM.RELIABLE_REPLY"
+
+
+def _extract_result(reply: Shareable, target: str):
+    if not isinstance(reply, dict):
+        return None, None
+    result = reply.get(target)
```

Comment:
If we hinted that reply is a Shareable, should we use shareable getter methods?
Someday we might change it to not inherit from dict?

OR we should just type hint it to dict?

Reply:
The reply is a dict, not shareable. Changed the hint.